Ericscript integration

tetianakh opened the pull request:
This pull request aims to integrate [EricScript](https://sites.google.com/site/bioericscript/) into bcbio (issue #1649 )

To run gene fusion detection with EricScript in rnaseq pipeline, you need to install the tool, download it's reference data, and add the `fusion_caller: ericscript` to the `algorithm` section of the config.

Changes:

1. Added a wrapper for EricScript, which runs the tool itself. EricScript takes 2 fastq files as an input, so there's not much logic in the wrapper. In case of an algorithm with disambiguation,  the .bam file with disambiguated reads is converted into 2 fastq files. Otherwise, the 2 fastq files with trimmed reads from the sample config object are used.
2. Slightly changed the logic of rnaseq pipeline. Previously, `fusion_mode` parameter was determining whether the fusion needs to be run, and the fusion calling was done with  an aligner of choice.
I've added a second option, `fusion_caller` (which can be left out), which can be used to specify a standalone caller like EricScript. If a `fusion_caller` is present, gene fusion detection won't be run with an aligner. My wild guess is that this is the most user-friendly behavior; however, this guess might be totally wrong, so we can change it back if you think it was a bad idea.
3.  Possibility to install EricScript with `bcbio upgrade` via `--toolplus` argument. Currently, EricScript requires `samtools=0.1.19`, which conflicts with bcbio's latest samtools verison. That's why EricScript is installed into its own conda environment. The EricScript's author plans to update the required `samtools` version, so whenever it's done, EricScript can be installed together with the rest of the tools or conda packages (EricScirpt and its dependencies are conda installable).
4. Possibility to install the latest version of [EricScript's database](https://sites.google.com/site/bioericscript/download). The reference data is currently shared via Google Drive, so I've added integration with Google Drive API to bcbio's `objectstore` module.
The tricky part is that [Google API requires API keys](https://developers.google.com/api-client-library/python/auth/service-accounts), even for publicly available files. I've made a set of keys, and put it in a .json file in the repo. You might want to get a set of keys, and maybe find a better way to store them.
5. As a consequence of integration with Google API, there's a new dependency on `google-api-client`. Currently those are listed in `requirements_dev.txt`, and also installed in `.travis.yml` in a hackish way. As soon as they're added to list of bcbio dependencies, we can of course remove them from those files.
6. Documentation updated (the installation section) with the instructions on how to install EricScript and its data.
7. A small change: renamed the `open` function from the `objectstore` module, because it was conflicting with python's built in `open`.
8. I didn't add any "full pipeline" tests; instead, I wrote 2 integration tests which run only the EricScript wrapper, and verify that the output is as expected. However, I've tested the full pipeline locally, and it works as intended. Sadly, I can't test all the possible parallelization scenarios, so a sanity check of the changes to the `bcbio/pipeline/main.py` is needed.

Any feedback is most welcome!

I am not sure whether I you prefer to keep commits small, squash them into a few large commits, or squash them into one ("one commit - one feature"). So currently there are just a lot of small commits. Please let me know if I should squash them somehow.

roryk commented on bcbio/install.py:
```diff
@@ -251,6 +256,7 @@ def get_genome_dir(gid, galaxy_dir, data):
 
 def _get_data_dir():
     base_dir = os.path.realpath(os.path.dirname(os.path.dirname(os.path.realpath(sys.executable))))
+    import ipdb; ipdb.set_trace()  # XXX BREAKPOINT
```
Looks like there is some stray debugging stuff.

roryk commented:
This is awesome. A couple things, I had to add __init__.py to the tests/unit/rnaseq and tests/integration/rnaseq to get py.test to behave ok. I'm having trouble installing ericscript via bioconda, how did you get it installed?

```
Fetching package metadata .........
Solving package specifications: .


PackageNotFoundError: Package not found: '' Dependencies missing in current linux-64 channels:
  - ericscript 0.5.5* -> r 3.3.1* -> r-base 3.3.1 -> bzip2
  - ericscript 0.5.5* -> r 3.3.1* -> r-base 3.3.1 -> icu 54.*
  - ericscript 0.5.5* -> r 3.3.1* -> r-base 3.3.1 -> libgcc
  - ericscript 0.5.5* -> r 3.3.1* -> r-base 3.3.1 -> jpeg 8d*

Close matches found; did you mean one of these?

    bzip2: pbzip2, zip, pbgzip
    libgcc: libgd, libmagic

You can search for packages on anaconda.org with

    anaconda search -t conda libgcc

(and similarly for the other packages)
```

chapmanb commented:
Tetiana;
Wow, this is incredible. Thank you for all the work integrating this. We're hoping to have a new release in the next couple of days and would love to pull this in afterwards, once Rory has had a chance to test and evaluate.

Rory, for the install you also need to include conda-forge and it looks like it'll pick up all the dependencies:
```
bcbio_conda install -c r -c conda-forge -c bioconda ericscript
```
Some thoughts:
- Is it possible to move the data installation to a CloudBioLinux ggd-recipe (https://github.com/chapmanb/cloudbiolinux/tree/master/ggd-recipes). The Google Drive logic is pretty complex, I wasn't sure if it's possible to replicate from the commandline. In general we've been trying to push installs as ggd-recipes so they could also be generally useful outside of bcbio, and more easily accessible to others looking at the logic.
- What is the source of the samtools 0.1.9 dependency? Ericscript is dependent on the command line, or something else? It would be awesome to get this as a standard install if we could resolve this and avoid needing a separate conda environment.

Thank you again for all the great work.

tetianakh commented:
Hello @chapmanb and @roryk, thank you for your kind words and your feedback.  I am sorry for the late reply, I was sick and out of the office last week.

@chapmanb 
> Is it possible to move the data installation to a CloudBioLinux ggd-recipe 

Theoretically, `ericscript` should be able to download its own databases, but the download is buggy.  I've mailed the author of the EricScript about a month ago, and he said that he plans to fix the issue. There's also a [post on their google group](https://groups.google.com/forum/#!topic/ericscript/__Zl4vPLwLI).

Downloading a file with the python API client seems unnecessarily complex for such a simple task.  But the problem with GoogleDrive is that one needs an API key to interact with it programmatically. It's not possible to get a file from the Drive "normally" (via `wget` or a similar command line utility). While it is possible in principle to download a file via plain HTTP requests, it's not clear to me how to run authentication. Additionally, the docs advise against this approach:

> the mechanics of server-to-server authentication interactions require applications to create and cryptographically sign JSON Web Tokens (JWTs), and it's easy to make serious errors ... For this reason, we strongly encourage you to use libraries, such as the Google APIs client libraries, that abstract the cryptography away from your application code. 

As the GoogleDrive is not among the most popular ways to store reference data, maybe there's no reason to keep this code in bcbcio as it won't be used much anyway. I can create a standalone python script which downloads a file from GoogleDrive and then can create a cloudbiolinux recipe using this script. 

Alternatively, we could ask users to download the file manually via browser until the issue is fixed.


> What is the source of the samtools 0.1.19 dependency? 

Samtools underwent some major changes after version 0.1.19, I guess EricScript hasn't adjusted to  newer versions yet. It simply fails with an error:
`Error: SAMtools >= 1.0 detected! EricScript is not yet compatible with it. Please use samtools 0.1.19 to run EricScript.`

On the bright side, a separate environment guarantees that the main bcbio installation won't be accidentally messed up by different versions of dependent packages. I wonder if you've encountered  dependency version conflicts in the past, and if so, how did you solve the problem? 


@roryk thanks a lot, I've removed the breakpoint, added the `conda-forge` channel, and added `__intit__.py` files.

chapmanb commented:
Tetiana;
Hope you're feeling better, and thank you again for all this work and explaining the approach. It sounds like some of these sticking points will be resolved over time but it makes sense to get this in now. If Rory is good with all the integration work I'm happy to merge this. Apologies but I created a couple of merge conflicts pulling in Michael's py3 work as well so after we squash those happy to get this all in and start testing it.

roryk commented:
I am psyched to have this in-- Brad, do the two remaining merge conflicts look like master handles them correctly?

chapmanb commented:
Rory -- it would be awesome if you want to fix the conflicts and get this merged. Happy to have this in place so you and others can start working with it.

roryk commented:
Yup, I'm on it.

biocyberman commented:
What's the status of this PR? I will be able to help with the merging if needed.

roryk commented:
Thanks Vang,

I had a heck of a time getting ericscript to be able to be installed by bioconda but it looks like it is installable now. This would be a nice fusion calling option to have integrated for sure.

biocyberman commented:
Nice work Rory. Wonder how the conflict of samtools version was solved. bcbio_nextgen currently uses `goleft` , which require samtools >=1.3 (I was hit by this conflict). One quick way I can think of is to alias samtools to fullpath of samtools 0.1.19 for `ericscript`. Or depending on how tightly `ericscript` integrates with samtools, we may be able to upgrade to samtools 1.3 for it.

Back to merge conflict, shall I try resolve conflict or you are in the process now?
Cheers
Vang

biocyberman commented:
Just did a test installation of ericscript:

```
/bcbio_nexgen/anaconda/bin/conda install ericscript 
Fetching package metadata .................                                                           
Solving package specifications: .                                                                     
                                                                                                      
Package plan for installation in environment /bcbio_nextgen/anaconda:                           
                                                                                                      
The following NEW packages will be INSTALLED:                                                         
                                                                                                      
    blat:               35-1                       bioconda                                           
    bleach:             1.4.2-py27_0               bioconda                                           
    configparser:       3.5.0-py27h5117587_0                                                          
    entrypoints:        0.2.3-py27h502b47d_2                                                          
    ericscript:         0.5.5-pl5.22.0r3.4.1_1     bioconda                                           
    html5lib:           0.999999999-py27hdf15f34_0                                                    
    ipywidgets:         7.0.0-py27h4fda95d_0                                                          
    mistune:            0.7.4-py27h6da7e90_0                                                          
    nbconvert:          5.3.1-py27he041f76_0                                                          
    notebook:           5.0.0-py27h3661c2b_2                                                          
    pandocfilters:      1.4.2-py27h428e1e5_1                                                          
    pytest-marks:       0.4-py27_0                 bioconda                                           
    r-ada:              2.0_5-r3.4.1_0             conda-forge                                        
    sh:                 1.11-py27_0                                                                   
    terminado:          0.6-py27h4be8df9_0                                                            
    testpath:           0.3.1-py27hc38d2c4_0                                                          
    webencodings:       0.5.1-py27hff10b21_1                                                          
    widgetsnbextension: 3.0.2-py27hcb77dec_1                                                          
                                                                                                      
The following packages will be UPDATED:                                                               
                                                                                                      
    conda:              4.3.29-py27_0              conda-forge --> 4.3.30-py27h6ae6dc7_0              
    conda-env:          2.6.0-0                    conda-forge --> 2.6.0-h36134e3_1                   
                                                                                                      
The following packages will be DOWNGRADED:                                                            
                                                                                                      
    bcbio-nextgen:      1.0.5-py27_0               bioconda    --> 1.0.3a-py27_6         bioconda     
    cnvkit:             0.8.6a0-py27_2             bioconda    --> 0.8.1-py27_0          bioconda     
    pysam:              0.11.2.2-py27_1            bioconda    --> 0.9.0-py27_1          bioconda     
    samtools:           1.6-0                      bioconda    --> 0.1.19-2              bioconda     
                                                                                                      
Proceed ([y]/n)? n                                                                                    
                                                                                                      
Exiting    
```                                                                                           
This means that samtools version conflicts still occur, and it is not preferable in my opinion.

chapmanb commented:
Vang and Rory;
Thanks so much for helping to get this in. The bcbio installation now has separate environments which we use for python3-only tools so I will work on creating an isolated install for eriscript that avoid the samtools conflict. With that in place we'd then need to remove the custom install code in this PR and clean up the conflicts. If either of you can work on those issues I'll tackle the ericscript install so we have something to test with. Thanks again for helping with this.

biocyberman commented:
@chapmanb @roryk @tetianakh I got some spare time on friday afternoon, so I played with git merge a bit and made this branch: https://github.com/biocyberman/bcbio-nextgen/tree/pr1724merged. **I haven't run any test, so use with precaution.**
My strategy is: rebase this PR on to 28c985f8, and then merge:
- prefer @tetianakh changes where you made improvement.
- parallel addition of new code are both kept
- prefer official version in `.travis.yml`

chapmanb commented:
Vang;
Thanks so much for working on this. I pushed the ericscript install to CloudBioLinux so doing `bcbio_nextgen.py upgrade --tools` will get it installed in a bcbio installation in an isolated conda environment. Looking through the PR the major things I'd like to do are:

- Back out custom conda install logic, in favor of fixes.
- Avoid changing travis and Docker files, we can address any dependency issues afterwards.
- Remove the changes to prun, since the logic there has changed slightly since the original PR.
- Remove the custom download and add this as a ggd-recipe. I can take on the data install.

Do you feel comfortable with git skills for updating and rebasing this? I've been nervous to tackle because I'm definitely going to mess that up but would be happy to get a non-conflicting branch in and tweak afterwards as well.

Thanks again for helping push this forward.

chapmanb commented:
Vang and Rory;
I pushed install targets for ericscript to CloudBioLinux as well so now both the tool and data installation are located there. We can remove the custom install code and only need a `ericscript` data target. This should make the PR a bit simpler by removing the custom Google integration so we can avoid all the extra dependency updates and focus on only the fusion, ericscript, disambiguation and testing changes. Thanks again for helping move this forward with us.

biocyberman commented:
Brad 
Cool with the additional `samtools0` conda env. I will take a closer look at the merge conflicts with the bulletpoints you mentioned.  I am comfortable with solving merge conflicts. Just wonder what data I can use to make a test run after merging.

biocyberman commented:
@chapmanb: This branch is already rebased and merged without conflict https://github.com/biocyberman/bcbio-nextgen/tree/pr1724merged. However, to revert/remove non-conflicting code it will take more time by going through all files of the PR. Would it be OK if I do merge without rebase, and then divide the merge code into several commits?

chapmanb commented:
Vang;
Thanks again for looking at this. I'm not too worried about rebase/merge on the branch so whatever works best practically is good for me. The history on this is already pretty messy so some more commits shouldn't make a big impact.

Practically we're aiming to do a 1.0.6 release soon and then will work on merging this after that is out, to avoid having a big set of code changes just prior to the release. Thanks again for all the help finalizing this.

chapmanb commented:
Thank you so much for this contribution. We've got it all cleaned up and pulled in PR #2146. I'm excited to have this finally sorted out and integrated and looking forward to testing and using it in bcbio. Our next aim will be to test and clean up before the next release. Thank you again.
